docs: DOC-1066: Vectorization guides

margaretkennedy opened the pull request.

margaretkennedy committed:
Vectorization guides

margaretkennedy committed:
.

margaretkennedy committed:
order string

elijahpetty commented on docs/python/sidebar.json:
```diff
@@ -63,6 +63,10 @@
                 "path": "getting-started/crash-course/query-strings.md"
               },
+              {
+                "label": "Vectorization",
```
Can we be more specific with the titles so we do not have two docs called Vectorization in different places?

elijahpetty commented on docs/python/conceptual/vectorization-and-recipes.md:
```diff
@@ -0,0 +1,655 @@
+title: Vectorization and the recipe paradigm
+---
+
+Deephaven's query engine uses vectorized operations and a declarative "recipe" paradigm to achieve high performance on both static and real-time data. This guide explains the technical foundations of this approach and why it matters for your queries.
```
There are two sections before the doc explains what the "recipe paradigm" is. I'd like a quick and simple definition somewhere at the beginning of the doc before the detailed explanation.

elijahpetty commented on docs/python/conceptual/vectorization-and-recipes.md:
````diff
@@ -0,0 +1,655 @@
+result = source.update_by(cum_sum("SumX = X"))
+```
+
+Watch this table in the UI. Every second:
````
Add a gif of this table so users don't have to run it to watch in the UI if they don't want to

elijahpetty commented on docs/python/getting-started/crash-course/vectorization-vs-loops.md:
```diff
@@ -0,0 +1,314 @@
+- How to chunk the data for optimal performance
+- Whether to parallelize the operation
+- How to handle updates efficiently
+- What rows need recomputation when data changes
```
periods?

elijahpetty reviewed (commented):
A few comments

margaretkennedy committed:
apply code review

margaretkennedy committed:
docs: refine vectorization guide examples and explanations

- Simplify SISD vs SIMD comparison by removing pandas and focusing on core Python concepts
- Remove performance timing example in favor of clearer conceptual explanations
- Update code examples to use executable test-set snippets instead of skip-test
- Clarify that loops require additional infrastructure for real-time updates rather than being "fundamentally impossible"
- Simplify technical explanations (e.g., ternary operator compilation

margaretkennedy commented:
I do not want to create the Groovy docs until the Python is reviewed and revised.

margaretkennedy committed:
.

margaretkennedy committed:
Merge branch 'main' into DOC-1066-looping

margaretkennedy committed:
fixes

cpwright commented on docs/groovy/conceptual/vectorization-and-recipes.md:
```diff
@@ -0,0 +1,572 @@
+
+### Deephaven: Declarative SIMD
+
+Deephaven uses a **declarative** approach. The engine processes data in chunks — Single Instruction, Multiple Data (SIMD) — applying one operation across many values at once:
```
We don't use SIMD instructions; we should not claim that we do.

margaretkennedy replied:
fixed

margaretkennedy committed:
Apply Charles's note

margaretkennedy committed:
snaps

margaretkennedy committed:
snap

margaretkennedy committed:
Fix malformed notes.

elijahpetty commented on docs/groovy/conceptual/vectorization-and-recipes.md:
```diff
@@ -0,0 +1,573 @@
+3. Engine automatically computes `Y` for the new rows.
+4. Updates propagate through the DAG.
+
+This requires significant additional infrastructure with imperative loops — a loop executes once and stops, so you would need to build your own subscription and recomputation logic.
```
Rephrase to avoid "would need" passive phrasing

margaretkennedy replied:
```suggestion
This requires significant additional infrastructure with imperative loops — because loops execute once and stop, you need to build your own subscription and recomputation logic.
```

elijahpetty commented on docs/groovy/conceptual/vectorization-and-recipes.md:
```diff
@@ -0,0 +1,573 @@
+- A new row arrives in `source`.
+- `XSquared` is computed for the new row.
+- `SumX` is updated for the new row.
+- **You wrote the recipe once, it runs forever**.
```
```suggestion

**Write the recipe once, and it runs forever**.
```

elijahpetty commented on docs/groovy/conceptual/vectorization-and-recipes.md:
````diff
@@ -0,0 +1,573 @@
+
+The recipe paradigm makes real-time processing trivial. Compare:
+
+#### Loop approach (doesn't work for real-time):
+
+```groovy skip-test
+// ❌ This only runs once!
+results = []
+source.columnIterator("X").forEachRemaining { x ->
+    results.add(x * 2)
+}
+// What happens when new data arrives? Nothing!
+```
+
+#### Recipe approach (automatically handles updates):
+
+```groovy skip-test
+// ✓ This updates automatically!
+result = source.update("Y = X * 2")
+// New data arrives? Y is computed for new rows automatically!
+```
````
These could be combined into one code block

elijahpetty commented on docs/groovy/conceptual/vectorization-and-recipes.md:
````diff
@@ -0,0 +1,573 @@
+```groovy ticking-table order=null test-set=groovy-live-agg
+import static io.deephaven.api.updateby.UpdateByOperation.RollingAvg
+
+// Streaming data with rolling statistics
````
```suggestion
// Streaming example data with rolling statistics
```

Or just "Example: Live aggregations" as a heading?

"Real-world" example seems misleading when we use fake data, but maybe that's just how I'm interpreting it.

elijahpetty commented on docs/groovy/conceptual/vectorization-and-recipes.md:
```diff
@@ -0,0 +1,573 @@
+
+See [`where`](../reference/table-operations/filter/where.md) for more details.
+
+Deephaven tables can share their RowSet with other tables in the same update graph that contain the same row keys. This sharing avoids copying data unnecessarily.
```
```suggestion
Deephaven tables can share their `RowSet` with other tables in the same update graph that contain the same row keys. This sharing avoids copying data unnecessarily.
```

elijahpetty reviewed (commented):
A few comments.

margaretkennedy commented on docs/groovy/conceptual/vectorization-and-recipes.md:
```diff
@@ -0,0 +1,573 @@
+- Dependencies are tracked automatically.
+- Downstream tables update accordingly.
+
+### Real-world example: Live aggregations
```
```suggestion
### Example: Live aggregations
```

margaretkennedy committed:
Apply suggestions from code review

Co-authored-by: elijahpetty <128415452+elijahpetty@users.noreply.github.com>
Co-authored-by: margaretkennedy <82049573+margaretkennedy@users.noreply.github.com>

margaretkennedy committed:
snaps

elijahpetty reviewed (approved):
LGTM
